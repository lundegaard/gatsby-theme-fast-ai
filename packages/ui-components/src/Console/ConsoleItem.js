import React, { memo } from 'react';
import PropTypes from 'prop-types';
// import { FormattedNumber } from 'gatsby-theme-fast-ai';
import { keyframes } from '@emotion/core';
import Dotdotdot from 'react-dotdotdot';

import Box from '../Box';
import Flex from '../Flex';

const backgroundFadeout = keyframes`
	0% {
		background-color: rgba(255, 0, 0, 1);
	}
	100% {
		background-color: rgba(255, 0, 0, 0);
	}
`;

const HighlightChangesText = ({ ...rest }) => (
	<Box
		{...rest}
		sx={{
			animationName: backgroundFadeout,
			animationDuration: '1.5s',
			animationTimingFunction: 'ease-out',
			animationFillMode: 'both',
		}}
	/>
);

const ConsoleItem = ({ value, label, ...rest }) => (
	<Flex sx={{ py: 2, px: 3, alignItems: 'center' }} {...rest}>
		<Box as="span" fontSize={1} mb={0}>
			{label}
		</Box>
		<HighlightChangesText title={value} key={value} as="span" ml="auto" mb={0} fontSize={1}>
			<Dotdotdot clamp={1}>{value}</Dotdotdot>
		</HighlightChangesText>
	</Flex>
);

ConsoleItem.propTypes = {
	label: PropTypes.string,
	value: PropTypes.any,
};

export default memo(ConsoleItem);
