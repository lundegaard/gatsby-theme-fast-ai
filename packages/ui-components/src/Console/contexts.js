import { createContext } from 'react';

export const ConsoleApiContext = createContext();
export const ConsoleLogContext = createContext([]);
