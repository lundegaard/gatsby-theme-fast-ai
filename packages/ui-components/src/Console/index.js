export { default } from './Console';
export { default as Console, defaultFormattingFunction } from './Console';
export { default as ConsoleProvider } from './ConsoleProvider';
export { default as ConsoleItem } from './ConsoleItem';

export * from './hooks';
