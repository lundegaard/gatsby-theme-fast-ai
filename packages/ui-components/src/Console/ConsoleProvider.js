import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import { ConsoleApiContext, ConsoleLogContext } from './contexts';

const ConsoleProvider = ({ children }) => {
	const [log, setLog] = useState({});

	const api = useMemo(
		() => ({
			log: (lines) => setLog((previousLog) => ({ ...previousLog, ...lines })),
			replace: (lines) => setLog(lines),
			clear: () => setLog({}),
		}),
		[]
	);

	return (
		<ConsoleApiContext.Provider value={api}>
			<ConsoleLogContext.Provider value={log}>
				{children}
			</ConsoleLogContext.Provider>
		</ConsoleApiContext.Provider>
	);
};
ConsoleProvider.propTypes = { children: PropTypes.node };

export default ConsoleProvider;
