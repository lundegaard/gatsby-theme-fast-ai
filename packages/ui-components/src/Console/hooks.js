import { useContext } from 'react';

import { ConsoleApiContext, ConsoleLogContext } from './contexts';

export const useConsole = () => useContext(ConsoleApiContext);
export const useConsoleLog = () => useContext(ConsoleLogContext);
