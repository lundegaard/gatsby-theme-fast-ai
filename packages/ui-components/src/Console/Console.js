import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import { keyframes } from '@emotion/core';
import { map, o, toPairs } from 'ramda';
import { isBoolean, isNumeric, isString } from 'ramda-extension';

import Box from '../Box';
import Flex from '../Flex';

import ConsoleItem from './ConsoleItem';
import { useConsoleLog } from './hooks';

const jumpInFromBottom = keyframes`
	0% {
		transform: translateY(200px) scaleY(2.5) scaleX(.2);
		transform-origin: 50% 0%;
		filter: blur(40px);
		opacity: 0;
	}
	100% {
		transform: translateY(0) scaleY(1) scaleX(1);
		transform-origin: 50% 50%;
		filter: blur(0);
		opacity: 1;
	}
}
`;

const easing = 'cubic-bezier(.455, .030, .515, .955)';

/* eslint-disable max-len */
const DropdownArrow = memo((props) => (
	<Box
		as="svg"
		width="13px"
		height="9px"
		viewBox="0 0 13 9"
		xmlns="http://www.w3.org/2000/svg"
		{...props}
	>
		<polygon
			fill="currentColor"
			points="10 1.03916875e-13 11.4061556 1.42222591 5.70307778 7.06086213 0 1.42222591 1.40615556 1.03916875e-13 5.70307778 4.24811295"
		/>
	</Box>
));
DropdownArrow.displayName = 'DropdownArrow';

export const defaultFormattingFunction = (x) => {
	if (isString(x)) {
		return x;
	}

	// if (isNumeric(x)) {
	// 	return <FormattedNumber value={x} maximumFractionDigits={1} />;
	// }

	if (isNumeric(x)) {
		return Math.round(x * 1000) / 1000;
	}

	if (isBoolean(x)) {
		return x ? 'True' : 'False';
	}

	return '-';
};
/* eslint-enable max-len */

const Console = ({
	title,
	formatValue = defaultFormattingFunction,
	initiallyOpened,
	...rest
}) => {
	const [isOpened, setIsOpened] = useState(initiallyOpened);
	const log = useConsoleLog();

	return (
		<Box
			sx={{
				position: 'fixed',
				right: '0',
				bottom: 0,
				zIndex: 99999,
				animationName: jumpInFromBottom,
				animationDuration: '.5s',
				animationTimingFunction: easing,
				animationFillMode: 'both',
				animationDelay: '.75s',
				backgroundColor: '#3B3B3B',
				color: 'background',
				maxHeight: '100vh',
				minWidth: ['auto', '400px'],
				width: ['100%', '33%'],
			}}
			{...rest}
		>
			<Flex
				sx={{ py: 2, px: 3, alignItems: 'center', cursor: 'pointer' }}
				onClick={() => setIsOpened(!isOpened)}
			>
				<Box as="span" fontFamily="mono" my={0} sx={{ pointerEvents: 'none' }}>
					{title}
				</Box>
				<DropdownArrow
					sx={{
						marginLeft: 'auto',
						display: 'inline',
						transform: `rotate(${isOpened ? '0' : '180'}deg)`,
					}}
				/>
			</Flex>

			{isOpened && (
				<Box
					sx={{
						borderTopColor: 'background',
						overflowY: 'auto',
						height: '300px',
						borderTopStyle: 'solid',
						borderTopWidth: '1px',
					}}
				>
					{log &&
						o(
							map(([label, value]) => (
								<ConsoleItem
									key={label}
									label={label}
									value={formatValue(value)}
								/>
							)),
							toPairs
						)(log)}
				</Box>
			)}
		</Box>
	);
};

Console.propTypes = {
	children: PropTypes.node,
	formatValue: PropTypes.func,
	initiallyOpened: PropTypes.bool,
	title: PropTypes.node,
};

export default Console;
