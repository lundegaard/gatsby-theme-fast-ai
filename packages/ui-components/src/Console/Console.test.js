import React from 'react';
import { shallow } from 'enzyme';

import Console from './Console';

describe('<Console />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Console {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
