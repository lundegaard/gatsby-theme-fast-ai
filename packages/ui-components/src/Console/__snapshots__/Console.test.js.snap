// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Console /> matches snapshot when shallowly rendered 1`] = `
<Box
  sx={
    {
      "animationDelay": ".75s",
      "animationDuration": ".5s",
      "animationFillMode": "both",
      "animationName": {
        "anim": 1,
        "name": "animation-1l35qkc",
        "styles": "@keyframes animation-1l35qkc{
	0% {
		transform: translateY(200px) scaleY(2.5) scaleX(.2);
		transform-origin: 50% 0%;
		filter: blur(40px);
		opacity: 0;
	}
	100% {
		transform: translateY(0) scaleY(1) scaleX(1);
		transform-origin: 50% 50%;
		filter: blur(0);
		opacity: 1;
	}
}
}",
        "toString": [Function],
      },
      "animationTimingFunction": "cubic-bezier(.455, .030, .515, .955)",
      "backgroundColor": "#3B3B3B",
      "bottom": 0,
      "color": "background",
      "maxHeight": "100vh",
      "minWidth": [
        "auto",
        "400px",
      ],
      "position": "fixed",
      "right": "0",
      "width": [
        "100%",
        "33%",
      ],
      "zIndex": 99999,
    }
  }
>
  <Flex
    onClick={[Function]}
    sx={
      {
        "alignItems": "center",
        "cursor": "pointer",
        "px": 3,
        "py": 2,
      }
    }
  >
    <Box
      as="span"
      fontFamily="mono"
      my={0}
      sx={
        {
          "pointerEvents": "none",
        }
      }
    />
    <DropdownArrow
      sx={
        {
          "display": "inline",
          "marginLeft": "auto",
          "transform": "rotate(180deg)",
        }
      }
    />
  </Flex>
</Box>
`;
